### ISSUE TITLE
cannot read from channel with job_start on windows

### ISSUE THREAD

**Issue by vrischmann** [vim/vim#1026]
Hi,

I've encountered a bug using job_start. When a program writes to stderr, the job randomly fails to send the output to the callback handler. Sometimes it does work and I get the output I expected.

This happens with the following script:

```
call ch_logfile('G:/debug_vim.log', 'w')

function! Handler(channel, msg)
    echom a:msg
endfunction

function! CloseHandler(ch)
    echom 'close:' . a:ch
endfunction

function! ExitHandler(job, exit_status)
  echom "exit: " . a:exit_status
endfunc

let command = [&shell, &shellcmdflag, "G:/debug.exe"]
let job = job_start(command, { "callback": "Handler", "exit_cb": "ExitHandler", "close_cb": "CloseHandler" })

echom job

call job_status(job)
```

I test the script either via `:source myscript.vim` or via `vim -S myscript.vim`.

`G:/debug.exe` is a Go program like this:

``` go
package main

import (
        "fmt"
        "os"
)

func main() {
        fmt.Fprintln(os.Stderr, "A")
        fmt.Fprintln(os.Stderr, "B")
}
```

I get this in my channel log file:

```
==== start log session ====
  0.000035 : Starting job: C:\WINDOWS\system32\cmd.exe /c G:/debug.exe
  0.000119 on 0: Created channel
  0.033090 : looking for messages on channels
  0.049383 ERR on 0: channel_handle_events()(): Cannot read from channel, will close it soon
  0.049405 : looking for messages on channels
  0.049409 on 0: Closing channel because of previous read error
  0.049412 on 0: Closing channel
  0.049428 on 0: Invoking callbacks before closing
  0.049431 on 0: Invoking close callback CloseHandler
  0.149783 : looking for messages on channels
 10.146457 on 0: Job ended
 10.246805 : looking for messages on channels
```

However, if I test the same program but make it write to stdout instead, it works correctly.

**Comment by vrischmann** [comment]
Ok so with these commands:

```
let command = [&shell, &shellcmdflag, "echo AB 1>&2"]
let command = [&shell, &shellcmdflag, "echo AB"]
```

I get the same results: first one doesn't work, second one always works. Might be easier to test that way

**Comment by vrischmann** [comment]
Did more digging tonight, and I noticed a couple more things, still no real fix though.

If I modify `channel_handle_events` to only check for the fd in PART_ERR, it works correctly.

I also noticed another thing for PART_OUT, PeekNamedPipe in `channel_wait` fails with ERROR_BROKEN_PIPE. Seems logical but in that case the whole channel is closed and therefore processing PART_ERR fails.

I tried removing [these lines](https://github.com/vim/vim/blob/master/src/channel.c#L3387-L3388) but it kept spamming my channel log file after the job was done (I did receive my output in the script though) so it's obviously a broken fix.

Basically right now I'm not sure if PeekNamedPipe is supposed to behave this way or if there's a logic flaw with channel_wait on Win32 ?

**Comment by trond-snekvik** [comment]
I have encountered this issue as well, and found that by sleeping in the start-thread until the close callback comes keeps this from happening (see example below). Since this blocks the GUI thread, it's not a proper work-around, but it could be a clue to a solution.

``` VimL
let s:in_progress = 0
fu! CloseHandler(channel)
    let s:in_progress = 0
endf

fu! StartJob()
    let s:in_progress = 1
    job_start(s:jobcommand, s:handlers)
    while s:in_progress == 1
        sleep 5m
    endwhile
endf
```

@vrischmann Have you made any progress on this?

**Comment by vrischmann** [comment]
@trond-snekvik no sorry, I haven't been able to find where the bug originates despite my digging. Up until now I just apply a patch to vim-go in order to make it work.

I'll have to recompile with the latest master and test again

**Comment by trond-snekvik** [comment]
Looks like this was fixed in 8.0.27, I'm not able to reproduce in 8.0.48.

**Comment by chrisbra** [comment]
closing then.
